### ISSUE TITLE
Rule Suggestion: PropsOnly option for no-param-reassign

### ISSUE THREAD

**Issue by jonathandelgado** [eslint/eslint#3814]
The current `no-param-reassign` rule allows for altering on props and variables being reassigned. This is great if you are worried about mutating the `arguments` variable.

If you aren't worried about that, however, if it still a good idea to look out for parameter object props being modified, since they alter variables outside of the function.

```javascript
var foo = function(obj){
  obj.prop = 2; //warn
  obj = 'str'; //no warn
}
```

If you are interesting in adding the option, I can throw in a PR.

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by ilyavolodin** [comment]
`no-param-reassign` will flag any assignment to arguments, so your second example will be flagged by default. Your first example will be flagged only when `props` is set to `true`.

I'm not sure I understand what you are asking. Could you provide more details please?

**Comment by jonathandelgado** [comment]
@ilyavolodin Sorry about that. I'm proposing an additional option to the `no-param-reassign` rule, which would allow argument reassignment, but not argument property reassignment. So the default behavior of the rule would effectively be ignored and only the `props` behavior would be observed.

The reasoning for this is that you may not be worried about the `arguments` variable being mutated, which the reason for the `no-param-reassign` rule, however, it is general practice to not modify properties of a parameter, since values are passed by reference. For example:

```
var process = function(obj, string){
  string = string.replace('y', 't');
  obj.test = string;
  console.log(obj);
}

var obj = {};
var string = 'yest';

process(obj, string);
console.log(obj.test); //test
console.log(string); //yest
```

From this example, you can see the object passed was modified, which caused the side-effect of the global object being change, since it was passed by reference. The string, however, was unchanged since javascript only passes arrays and objects by reference.

In summary, add an option to ignore the default behavior and only alert on prop modifications.

**Comment by mysticatea** [comment]
@jonathandelgado To make sure, how do you think the following patterns?

```js
function foo(obj) {
    obj = Object.assign({}, obj);
    obj.bar = 1; // should this be warned?
}

function foo(obj) {
    if (obj.abc) {
        obj = Object.assign({}, obj);
        obj.bar = 1; // should this be warned?
    } else {
        obj.bar = 2; // should this be warned?
    }
}
```

**Comment by jonathandelgado** [comment]
@mysticatea Exactly what I am trying to match.

**Comment by ilyavolodin** [comment]
I think in order to provide such an option, we would need to rename this rule.

**Comment by jonathandelgado** [comment]
Oh ok, will close then, no need to break backwards comparability for this.

**Comment by Erid** [comment]
A new rule for this sounds like the way to go, has one already been suggested?. If not I think `no-reassign-param-props` would do... It's basically the same but only for the props.

Other option would be to extend the rule and add {"root": true} as default (or similar), though the rule could easily mean nothing with both props set to false.

--

I believe this rule is specially useful because it's common that we don't want to mutate the reference of the object, but sometimes creating a new variable just for a new assignment is a little more verbose.
```
function foo(obj) {
  var newObj = obj
  if (!obj.get('a'))  newObj = obj.set()
  return next(newObj)
}
```
vs
```
function foo(obj) {
  if (!obj.get('a'))  obj = obj.set()
  return next(obj)
}
```
